--- drivers/usb/gadget/ether.c.bak	2010-03-31 13:51:37.000000000 +0800
+++ drivers/usb/gadget/ether.c	2010-03-31 13:51:18.000000000 +0800
@@ -263,6 +263,12 @@
 #define DEV_CONFIG_CDC
 #endif
 
+/* plany@20100326 */
+#ifdef CONFIG_USB_GADGET_STR8100
+#define DEV_CONFIG_CDC
+#endif
+
+
 /* For CDC-incapable hardware, choose the simple cdc subset.
  * Anything that talks bulk (without notable bugs) can do this.
  */
@@ -1561,6 +1567,7 @@
 				memcpy(req->buf, buf, n);
 				req->complete = rndis_response_complete;
 				rndis_free_response(dev->rndis_config, buf);
+				value = n;
 			}
 			/* else stalls ... spec says to avoid that */
 		}
